src/server/server.cc

```diff
@@ -1637,6 +1639,95 @@ Status ServerLogData::Decode(const rocksdb::Slice &blob) {
     content_ = std::string(blob.data() + 2, blob.size() - 2);
     return Status::OK();
   }
-
   return {Status::NotOK};
 }
+
+void Server::updateWatchedKeysFromRange(const std::vector<std::string> &args, const Redis::CommandKeyRange &range) {
+  std::shared_lock lock(watched_key_mutex_);
+
+  for (size_t i = range.first_key; range.last_key > 0 ? i <= size_t(range.last_key) : i <= args.size() + range.last_key;
+       i += range.key_step) {
+    if (auto iter = watched_key_map_.find(args[i]); iter != watched_key_map_.end()) {
+      for (auto *conn : iter->second) {
+        conn->watched_keys_modified_ = true;
+      }
+    }
+  }
+}
+
+void Server::updateAllWatchedKeys() {
+  std::shared_lock lock(watched_key_mutex_);
+
+  for (auto &[_, conn_map] : watched_key_map_) {
+    for (auto *conn : conn_map) {
+      conn->watched_keys_modified_ = true;
+    }
+  }
+}
+
+void Server::UpdateWatchedKeysFromArgs(const std::vector<std::string> &args, const Redis::CommandAttributes &attr) {
+  if (attr.is_write() && watched_key_size_ > 0) {
+    if (attr.key_range.first_key > 0) {
+      updateWatchedKeysFromRange(args, attr.key_range);
+    } else if (attr.key_range.first_key < 0) {
+      Redis::CommandKeyRange range = attr.key_range_gen(args);
+
+      if (range.first_key > 0) {
+        updateWatchedKeysFromRange(args, range);
+      }
```

Comment:
after `attr.key_range_gen`, if `range.first_key == 0`, would it be a command like flushdb?